test/controllers/tasks_controller_test.rb

```diff
@@ -71,56 +64,240 @@
           completion_date: nil,
         },
       }
-
+      
       # Act-Assert
       expect {
         post tasks_path, params: task_hash
       }.must_change "Task.count", 1
-
+      
       new_task = Task.find_by(name: task_hash[:task][:name])
       expect(new_task.description).must_equal task_hash[:task][:description]
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
-      expect(new_task.completed).must_equal task_hash[:task][:completed]
-
+      
       must_respond_with :redirect
       must_redirect_to task_path(new_task.id)
     end
+    
+    it "redirect to new_task page if input is not valid" do
+      task_hashes = [
+        {
+          task: {
+            name: "No description",
+            description: nil,
+            completion_date: nil,
+          },
+        },
+        {
+          task: {
+            name: nil,
+            description: "No name task description",
+            completion_date: nil,
+          },
+        },
+        {
+          task: {
+            name: "",
+            description: "      ",
+            completion_date: nil,
+          }
+        },
+        {
+          task: {
+            
+          }
+        }
+      ]
+      
+      task_hashes.each do |task_hash|
+        expect {
+          post tasks_path, params: task_hash
+        }.must_differ "Task.count", 0
+        
+        must_respond_with :redirect
+        must_redirect_to new_task_path
+      end
+    end
   end
-
-  # Unskip and complete these tests for Wave 3
+  
+  # Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      get edit_task_path(task[:id])
+      
+      must_respond_with :success
     end
-
+    
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      get edit_task_path(-1)
+      
+      must_respond_with :redirect
     end
   end
-
-  # Uncomment and complete these tests for Wave 3
+  
+  # Wave 3
   describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
-    it "can update an existing task" do
-      # Your code here
+    before do 
+      @task_hash = {
+        task: {
+          name: "updated task",
+          description: "updated task description", 
+          completion_date: nil
+        }
+      }
     end
 
+    it "can update an existing task" do
+      existing_task = Task.find_by(id: task[:id])
+      expect _(Task.count).must_equal 1
```

Comment:
this is super interesting syntax; I actually don't know what this is! I'll probably ask you at some point, since I haven't seen this before.

Reply:
@tildeee  Oh this is one of the weird things that minitest suggested me to do. Instead of ``` expect (obj).must_equal value``` minitest complains that this should be replaced with ``` expect _(obj).must_equal value``` so that the test won't fail on minitest 6. I had too many warnings about this I couldn't see where the stack trace was for my errors so I had to change my tests to use this syntax.